File: php-app/api/relatorios.php
<?php
// ==========================================
// API REST - RELATÓRIOS
// ==========================================
require_once '../config/database.php';
setJsonHeaders();

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';

try {
    $db = getDB();
    
    if ($method !== 'GET') {
        jsonResponse(['error' => 'Método não permitido'], 405);
    }
    
    if ($action === 'mensal' && isset($_GET['mes']) && isset($_GET['ano'])) {
        // Relatório mensal
        $mes = (int)$_GET['mes'];
        $ano = (int)$_GET['ano'];
        
        // Buscar imóveis ativos no período
        $stmt = $db->prepare("
            SELECT i.*, c.nome as cliente_nome, c.cpf as cliente_cpf
            FROM imoveis i
            LEFT JOIN clientes c ON i.cliente_uuid = c.uuid
            WHERE i.ativo = 1
            AND DATE_ADD(i.data_inicio, INTERVAL i.contrato_meses MONTH) >= ?
            AND i.data_inicio <= ?
            ORDER BY i.endereco
        ");
        $dataInicio = "$ano-" . str_pad($mes, 2, '0', STR_PAD_LEFT) . "-01";
        $dataFim = "$ano-" . str_pad($mes, 2, '0', STR_PAD_LEFT) . "-31";
        $stmt->execute([$dataInicio, $dataFim]);
        $imoveis = $stmt->fetchAll();
        
        // Buscar pagamentos do mês
        $stmt = $db->prepare("
            SELECT * FROM pagamentos
            WHERE mes = ? AND ano = ?
        ");
        $stmt->execute([$mes, $ano]);
        $pagamentos = $stmt->fetchAll();
        
        // Criar mapa de pagamentos
        $pagamentosMap = [];
        foreach ($pagamentos as $p) {
            $pagamentosMap[$p['imovel_uuid']] = $p;
        }
        
        // Calcular totais
        $totalReceita = 0;
        $totalRecebido = 0;
        $totalPendente = 0;
        $imoveisPagos = 0;
        $imoveisPendentes = 0;
        
        foreach ($imoveis as &$imovel) {
            $totalReceita += $imovel['valor_aluguel'];
            
            if (isset($pagamentosMap[$imovel['uuid']]) && $pagamentosMap[$imovel['uuid']]['pago']) {
                $imovel['pago'] = true;
                $imovel['data_pagamento'] = $pagamentosMap[$imovel['uuid']]['data_pagamento'];
                $totalRecebido += $imovel['valor_aluguel'];
                $imoveisPagos++;
            } else {
                $imovel['pago'] = false;
                $totalPendente += $imovel['valor_aluguel'];
                $imoveisPendentes++;
            }
        }
        
        jsonResponse([
            'mes' => $mes,
            'ano' => $ano,
            'total_receita' => $totalReceita,
            'total_recebido' => $totalRecebido,
            'total_pendente' => $totalPendente,
            'imoveis_pagos' => $imoveisPagos,
            'imoveis_pendentes' => $imoveisPendentes,
            'total_imoveis' => count($imoveis),
            'imoveis' => $imoveis
        ]);
        
    } elseif ($action === 'anual' && isset($_GET['ano'])) {
        // Relatório anual
        $ano = (int)$_GET['ano'];
        
        $meses = [];
        $totalAnual = 0;
        $totalRecebidoAnual = 0;
        
        for ($mes = 1; $mes <= 12; $mes++) {
            // Buscar imóveis ativos no mês
            $stmt = $db->prepare("
                SELECT i.valor_aluguel
                FROM imoveis i
                WHERE i.ativo = 1
                AND DATE_ADD(i.data_inicio, INTERVAL i.contrato_meses MONTH) >= ?
                AND i.data_inicio <= ?
            ");
            $dataInicio = "$ano-" . str_pad($mes, 2, '0', STR_PAD_LEFT) . "-01";
            $dataFim = "$ano-" . str_pad($mes, 2, '0', STR_PAD_LEFT) . "-31";
            $stmt->execute([$dataInicio, $dataFim]);
            $imoveis = $stmt->fetchAll();
            
            $totalMes = array_sum(array_column($imoveis, 'valor_aluguel'));
            
            // Buscar pagamentos do mês
            $stmt = $db->prepare("
                SELECT p.*, i.valor_aluguel
                FROM pagamentos p
                LEFT JOIN imoveis i ON p.imovel_uuid = i.uuid
                WHERE p.mes = ? AND p.ano = ? AND p.pago = 1
            ");
            $stmt->execute([$mes, $ano]);
            $pagos = $stmt->fetchAll();
            
            $recebidoMes = array_sum(array_column($pagos, 'valor_aluguel'));
            
            $meses[] = [
                'mes' => $mes,
                'nome' => ['Janeiro', 'Fevereiro', 'Março', 'Abril', 'Maio', 'Junho', 'Julho', 'Agosto', 'Setembro', 'Outubro', 'Novembro', 'Dezembro'][$mes - 1],
                'total' => $totalMes,
                'recebido' => $recebidoMes,
                'pendente' => $totalMes - $recebidoMes,
                'qtd_imoveis' => count($imoveis),
                'qtd_pagos' => count($pagos)
            ];
            
            $totalAnual += $totalMes;
            $totalRecebidoAnual += $recebidoMes;
        }
        
        jsonResponse([
            'ano' => $ano,
            'total_anual' => $totalAnual,
            'total_recebido' => $totalRecebidoAnual,
            'total_pendente' => $totalAnual - $totalRecebidoAnual,
            'meses' => $meses
        ]);
        
    } elseif ($action === 'stats') {
        // Estatísticas gerais
        $stmt = $db->prepare("SELECT COUNT(*) as total FROM clientes WHERE ativo = 1");
        $stmt->execute();
        $totalClientes = $stmt->fetch()['total'];
        
        $stmt = $db->prepare("SELECT COUNT(*) as total FROM imoveis WHERE ativo = 1");
        $stmt->execute();
        $totalImoveis = $stmt->fetch()['total'];
        
        $stmt = $db->prepare("
            SELECT COUNT(*) as total FROM pagamentos WHERE pago = 1
        ");
        $stmt->execute();
        $totalPagos = $stmt->fetch()['total'];
        
        $stmt = $db->prepare("SELECT COUNT(*) as total FROM pagamentos");
        $stmt->execute();
        $totalPagamentos = $stmt->fetch()['total'];
        
        $stmt = $db->prepare("
            SELECT SUM(valor_aluguel) as total
            FROM imoveis
            WHERE ativo = 1
            AND DATE_ADD(data_inicio, INTERVAL contrato_meses MONTH) >= CURDATE()
        ");
        $stmt->execute();
        $receitaMensal = $stmt->fetch()['total'] ?? 0;
        
        jsonResponse([
            'total_clientes' => $totalClientes,
            'total_imoveis' => $totalImoveis,
            'total_pagamentos' => $totalPagamentos,
            'total_pagos' => $totalPagos,
            'receita_mensal' => $receitaMensal
        ]);
        
    } else {
        jsonResponse(['error' => 'Ação inválida ou parâmetros faltando'], 400);
    }
    
} catch (Exception $e) {
    jsonResponse(['error' => $e->getMessage()], 500);
}
?>

File: python-app/routes/relatorios.py
from flask import Blueprint, request, jsonify
from models import Cliente, Imovel, Pagamento
from app import db
from datetime import datetime, date
from dateutil.relativedelta import relativedelta

relatorios_bp = Blueprint('relatorios', __name__)

def is_contrato_ativo_no_periodo(imovel, mes, ano):
    """Verifica se o contrato está ativo no mês/ano especificado"""
    data_inicio_mes = date(ano, mes, 1)
    if mes == 12:
        data_fim_mes = date(ano + 1, 1, 1) - relativedelta(days=1)
    else:
        data_fim_mes = date(ano, mes + 1, 1) - relativedelta(days=1)
    
    data_fim_contrato = imovel.data_inicio + relativedelta(months=imovel.contrato_meses)
    
    return data_inicio_mes <= data_fim_contrato and data_fim_mes >= imovel.data_inicio

@relatorios_bp.route('/stats', methods=['GET'])
def stats():
    """Estatísticas gerais"""
    total_clientes = Cliente.query.filter_by(ativo=True).count()
    total_imoveis = Imovel.query.filter_by(ativo=True).count()
    total_pagamentos = Pagamento.query.count()
    total_pagos = Pagamento.query.filter_by(pago=True).count()
    
    # Receita mensal (imóveis ativos hoje)
    hoje = date.today()
    imoveis_ativos = Imovel.query.filter_by(ativo=True).all()
    receita_mensal = sum([
        imovel.valor_aluguel for imovel in imoveis_ativos
        if is_contrato_ativo_no_periodo(imovel, hoje.month, hoje.year)
    ])
    
    return jsonify({
        'total_clientes': total_clientes,
        'total_imoveis': total_imoveis,
        'total_pagamentos': total_pagamentos,
        'total_pagos': total_pagos,
        'receita_mensal': receita_mensal
    })

@relatorios_bp.route('/mensal', methods=['GET'])
def relatorio_mensal():
    """Relatório mensal"""
    mes = request.args.get('mes', type=int)
    ano = request.args.get('ano', type=int)
    
    if not mes or not ano:
        return jsonify({'error': 'Parâmetros obrigatórios: mes, ano'}), 400
    
    # Buscar imóveis ativos no período
    imoveis = Imovel.query.filter_by(ativo=True).all()
    imoveis_ativos = [
        imovel for imovel in imoveis
        if is_contrato_ativo_no_periodo(imovel, mes, ano)
    ]
    
    # Buscar pagamentos do mês
    pagamentos = Pagamento.query.filter_by(mes=mes, ano=ano).all()
    pagamentos_map = {p.imovel_uuid: p for p in pagamentos}
    
    # Calcular totais
    total_receita = 0
    total_recebido = 0
    imoveis_dados = []
    
    for imovel in imoveis_ativos:
        total_receita += imovel.valor_aluguel
        
        pagamento = pagamentos_map.get(imovel.uuid)
        pago = pagamento.pago if pagamento else False
        
        if pago:
            total_recebido += imovel.valor_aluguel
        
        imoveis_dados.append({
            'uuid': imovel.uuid,
            'endereco': imovel.endereco,
            'tipo': imovel.tipo,
            'valor_aluguel': imovel.valor_aluguel,
            'cliente_nome': imovel.cliente.nome if imovel.cliente else None,
            'cliente_cpf': imovel.cliente.cpf if imovel.cliente else None,
            'pago': pago,
            'data_pagamento': pagamento.data_pagamento.isoformat() if pagamento and pagamento.data_pagamento else None
        })
    
    return jsonify({
        'mes': mes,
        'ano': ano,
        'total_receita': total_receita,
        'total_recebido': total_recebido,
        'total_pendente': total_receita - total_recebido,
        'imoveis_pagos': sum(1 for i in imoveis_dados if i['pago']),
        'imoveis_pendentes': sum(1 for i in imoveis_dados if not i['pago']),
        'total_imoveis': len(imoveis_ativos),
        'imoveis': imoveis_dados
    })

@relatorios_bp.route('/anual', methods=['GET'])
def relatorio_anual():
    """Relatório anual"""
    ano = request.args.get('ano', type=int)
    
    if not ano:
        return jsonify({'error': 'Parâmetro obrigatório: ano'}), 400
    
    meses_nomes = ['Janeiro', 'Fevereiro', 'Março', 'Abril', 'Maio', 'Junho',
                   'Julho', 'Agosto', 'Setembro', 'Outubro', 'Novembro', 'Dezembro']
    
    meses_dados = []
    total_anual = 0
    total_recebido_anual = 0
    
    for mes in range(1, 13):
        # Buscar imóveis ativos no mês
        imoveis = Imovel.query.filter_by(ativo=True).all()
        imoveis_ativos = [
            imovel for imovel in imoveis
            if is_contrato_ativo_no_periodo(imovel, mes, ano)
        ]
        
        total_mes = sum(imovel.valor_aluguel for imovel in imoveis_ativos)
        
        # Buscar pagamentos do mês
        pagamentos = Pagamento.query.filter_by(mes=mes, ano=ano, pago=True).all()
        recebido_mes = sum(
            imovel.valor_aluguel for imovel in imoveis_ativos
            if any(p.imovel_uuid == imovel.uuid for p in pagamentos)
        )
        
        meses_dados.append({
            'mes': mes,
            'nome': meses_nomes[mes - 1],
            'total': total_mes,
            'recebido': recebido_mes,
            'pendente': total_mes - recebido_mes,
            'qtd_imoveis': len(imoveis_ativos),
            'qtd_pagos': len(pagamentos)
        })
        
        total_anual += total_mes
        total_recebido_anual += recebido_mes
    
    return jsonify({
        'ano': ano,
        'total_anual': total_anual,
        'total_recebido': total_recebido_anual,
        'total_pendente': total_anual - total_recebido_anual,
        'meses': meses_dados
    })
